Separate out the Backbone image field

andy-armstrong opened the pull request:
I've broken out the Backbone image field so that views can use all the other fields without depending upon the file uploaded and hence JQuery UI. This decreased the size of the optimized teams_tab_factory from 430 KB to 125 KB.

andy-armstrong commented:
@cahrens @symbolist While profiling teams, we noticed that JQuery UI was being loaded unnecessarily. This change separates out the ImageField into its own Backbone view file, and only the learner profile page needs to load it. While I was here I cleaned up the RequireJS text usage which was a little inconsistent.

cahrens commented on lms/static/js/spec/student_profile/learner_profile_view_spec.js:
```diff
@@ -123,12 +123,7 @@ define(['backbone', 'jquery', 'underscore', 'common/js/spec_helpers/ajax_helpers
-                TemplateHelpers.installTemplate('templates/fields/field_dropdown');
-                TemplateHelpers.installTemplate('templates/fields/field_textarea');
-                TemplateHelpers.installTemplate('templates/fields/field_image');
                 TemplateHelpers.installTemplate('templates/fields/message_banner');
```
Why did you decide not to convert message_banner?

andy-armstrong replied:
It broke the tests when I took it out. I'll take a quick pass through and work out why.

andy-armstrong replied:
I remembered the problem. The message banner is used by some non-optimized code as well (donations) so I was nervous about changing it to use RequireJS Text. Thinking it through some more, I think it will be okay as it will dynamically load the template (without caching) until that code is converted to use Optimizer too.

cahrens commented on lms/static/js/views/image_field.js:
```diff
@@ -0,0 +1,230 @@
+        'gettext', 'jquery', 'underscore', 'backbone', 'js/views/fields',
+        'text!templates/fields/field_image.underscore',
+        'backbone-super', 'jquery.fileupload'
+    ], function (gettext, $, _, Backbone, FieldViews, field_image_template) {
```
None of this code was changed, correct? It's simply a move of code out of fields.js.

andy-armstrong replied:
Right, I didn't make any changes.

andy-armstrong commented:
@cahrens I've converted MessageBanner to use RequireJS Text, and I've also removed all of the static templates that were being included in the Mako templates. I'm going to run these changes on a sandbox and see how much of a difference they make.

andy-armstrong commented:
jenkins run bokchoy

andy-armstrong commented:
@cahrens @symbolist @muhammad-ammar @muzaffaryousaf I've updated my sandbox (http://andy-armstrong.sandbox.edx.org/)  and ran comparisons with it and production. This change decreases the size of the initial document for both account settings and learner profile because the templates are no longer statically included (account settings is 19K down from 48K). It also reduces the size of account_settings_factory.js from 265 KB to 28 KB by removing JQuery UI and File Image Upload. 

Unfortunately learner_profile_factory.js is unaffected because it needs both of those libraries, so it i still 275 KB. We should investigate using the non-JQuery UI version of File Image Upload since we aren't using the UI pieces. It looks like we need to upgrade to a newer version to be able to do this:

https://blueimp.github.io/jQuery-File-Upload/basic-plus.html

We might also consider skipping a plugin entirely and just implementing it ourselves:

http://stackoverflow.com/questions/19920322/how-to-upload-file-through-jquery-ajax

muzaffaryousaf commented:
@andy-armstrong Awesome. :+1: for reducing the page sizes. Great job. :trophy:

cahrens commented:
:+1: Great work, @andy-armstrong. I'm loving RequireJS text.

Please squash commits before merging.
